api/objects.proto

```diff
@@ -222,3 +222,54 @@ message Cluster {
 	// a new key is allocated on key rotation.
 	uint64 encryption_key_lamport_clock = 6;
 }
+
+// Secret represents a conceptual secret that should be passed to a container.
+// It can contain many versions of secret data, since conceptually secret
+// credentials can be rotated.
+message Secret {
+	string id = 1 [(gogoproto.customname) = "ID"];
+
+	Meta meta = 2 [(gogoproto.nullable) = false];
+
+	// A secret has multiple versions, and each version is represented by a
+	// `SecretData` object and referenced by the `SecretData`'s ID.  This map
+	// stores `SecretData` objects keyed by ID, which lets us reference any
+	// particular version of a secret easily.
+	map<string, SecretData> secret_data = 3;
+
+	// Name of the secret, so that we can use it easily for lookups.  This is
+	// just a copy of the data in `SecretData.Spec.Annotations.Name`.
+	string name = 4;
+
+	// Latest is the cached ID of the latest version, so that we can easily use
+	// it for lookups without having to sort the `SecretData` map, although
+	// we will have to sort and get the latest version if the `LatestVersion`
+	// is removed.
+	string latest = 5 [(gogoproto.customname) = "LatestVersion"];
+}
+
+// SecretData is represents a specific version of a secret, and is immutable;
+// it wraps the `spec` provided by the user with useful information that is
+// generated from the secret data in the `spec`, such as the digest and size
+// of the secret data.
+message SecretData {
+	string id = 1 [(gogoproto.customname) = "ID"];
```

Comment:
Rather than having identifiers for each revision of secret data, it might be better to use the digest and have a linked list to the previous version of the secret data.